File: 2015/day_1/juliaimp/solve.jl

problem_text = open("data/input.txt","r") do f
    read(f, String)
end

function part_1(text)
    floor = 0
    for c in strip(text)
        if c == '('
            floor += 1
        elseif c == ')'
            floor -= 1
        else
            error("unknown character: $c")
        end
    end
    return  floor
end

function part_2(text)
    floor = 0
    step = 0
    for (i, c) in enumerate(strip(text))
        if c == '('
            floor += 1
        elseif c == ')'
            floor -= 1
        else
            error("unknown character: $c")
        end
        if floor == -1
            step = i
            break
        end
    end
    return  step
end

p1 = part_1(problem_text)
println("part_1 = $p1")

p2 = part_2(problem_text)
println("part_2 = $p2")

File: 2015/day_03/pythonimp/solve.py
def load_text(path):
    """
    >>> text = load_text("../data/input.txt")
    >>> len(text.strip())
    8192
    """
    with open(path) as f:
        return f.read()


def solve_1(text):
    """
    >>> solve_1(">")
    2
    >>> solve_1("^>v<")
    4
    >>> solve_1("^v^v^v^v^v")
    2
    >>> solve_1(load_text("../data/input.txt"))
    2565


    >>> 11 % 5
    >>> 10 % 5
    """
    board = set([(0, 0)])
    i = 0
    j = 0
    for c in text:
        match c:
            case ">":
                j += 1
            case "<":
                j -= 1
            case "^":
                i -= 1
            case "v":
                i += 1
            case _:
                raise ValueError(f"unexpected symbol: '{c}'")
        board.add((i, j))
    return len(board)


if __name__ == "__main__":
    import doctest

    doctest.testmod()
